exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/scan/file/FileMetadataColumnsParser.java

```diff
@@ -0,0 +1,166 @@
+    if (inCol.isWildcard()) {
+      buildWildcard();
+
+      // Don't consider this a match.
+    }
+    return false;
+  }
+
+  private boolean buildPartitionColumn(Matcher m, RequestedColumn inCol) {
+
+    // If the projected column is a map or array, then it shadows the
+    // partition column. Example: dir0.x, dir0[2].
+
+    if (! inCol.isSimple()) {
+      logger.warn("Partition column {} is shadowed by a projected {}",
+          inCol.name(), inCol.summary());
+      return false;
+    }
+    if (builder.hasWildcard()) {
+      wildcardAndMetadataError();
+    }
+
+    // Partition column
+
+    builder.addMetadataColumn(
+        new PartitionColumn(
+          inCol.name(),
+          Integer.parseInt(m.group(1))));
+    hasMetadata = true;
+    return true;
+  }
+
+  private boolean buildMetadataColumn(FileMetadataColumnDefn defn,
+      RequestedColumn inCol) {
+
+    // If the projected column is a map or array, then it shadows the
+    // metadata column. Example: filename.x, filename[2].
+
+    if (! inCol.isSimple()) {
+      logger.warn("File metadata column {} is shadowed by a projected {}",
+          inCol.name(), inCol.summary());
+      return false;
+    }
+    if (builder.hasWildcard()) {
+      wildcardAndMetadataError();
+    }
+
+    // File metadata (implicit) column
+
+    builder.addMetadataColumn(new FileMetadataColumn(inCol.name(), defn));
+    hasMetadata = true;
+    return true;
+  }
+
+  private void buildWildcard() {
+    if (hasMetadata) {
+      wildcardAndMetadataError();
+    }
+    if (metadataManager.useLegacyWildcardExpansion) {
+
+      // Star column: this is a SELECT * query.
+
+      // Old-style wildcard handling inserts all metadata columns in
+      // the scanner, removes them in Project.
+      // Fill in the file metadata columns. Can do here because the
+      // set is constant across all files.
+
+      expandWildcard();
+      hasMetadata = true;
+    }
+  }
+
+  protected void expandWildcard() {
+
+    // Legacy wildcard expansion: include the file metadata and
+    // file partitions for this file.
+    // This is a disadvantage for a * query: files at different directory
+    // levels will have different numbers of columns. Would be better to
+    // return this data as an array at some point.
+    // Append this after the *, keeping the * for later expansion.
+
+    for (FileMetadataColumnDefn iCol : metadataManager.fileMetadataColDefns()) {
+      builder.addMetadataColumn(new FileMetadataColumn(
+          iCol.colName(), iCol));
+    }
+    for (int i = 0; i < metadataManager.partitionCount(); i++) {
+      builder.addMetadataColumn(new PartitionColumn(
+          metadataManager.partitionName(i), i));
+    }
+  }
+
+  @Override
+  public void validate() { }
+
+  @Override
+  public void validateColumn(ColumnProjection outCol) { }
+
+  private void wildcardAndMetadataError() {
+    throw new IllegalArgumentException("Cannot select file metadata columns and `*` together");
```

Comment:
I have added implicit columns support in April 2016, there were no limitations from the Calcite side on such syntax.  Since such cases are allowed and work now, I guess they need to work after your changes as well.